stlab/concurrency/actor.hpp

```diff
@@ -0,0 +1,179 @@
+        _q(
+            [_this = this->shared_from_this()](auto&&... args) mutable {
+                temp_thread_name ttn(_this->_name.c_str());
+                _this->_instance._x = T(std::forward<Args>(args)...);
+            },
+            std::forward<Args>(args)...)
+            .detach();
+    }
+
+    auto set_name(std::string&& name) { _name = std::move(name); }
+
+    template <typename F, typename... Args>
+    auto send(F&& f, Args&&... args) {
+        auto task = [_f = std::forward<F>(f),
+                     _this = this->shared_from_this()](auto&&... args) mutable {
+            temp_thread_name ttn(_this->_name.c_str());
+            if constexpr (std::is_same_v<T, void>) {
+                return std::move(_f)(std::forward<Args>(args)...);
+            } else {
+                return std::move(_f)(*(_this->_instance._x), std::forward<Args>(args)...);
+            }
+        };
+        return _q(std::move(task), std::forward<Args>(args)...);
+    }
+
+    template <typename R, typename F, typename... Args>
+    auto then(stlab::future<R>&& future, F&& f, Args&&... args) {
+        auto task = [_f = std::forward<F>(f), ... _args = std::forward<Args>(args),
+                     _this = this->shared_from_this()](R&& x) mutable {
+            temp_thread_name ttn(_this->_name.c_str());
+            if constexpr (std::is_same_v<T, void>) {
+                return std::move(_f)(std::forward<R>(x), std::move(_args)...);
+            } else {
+                return std::move(_f)(*(_this->_instance._x), std::forward<R>(x),
+                                     std::move(_args)...);
+            }
+        };
+        return std::move(future).then(_q.executor(), std::move(task));
+    }
+
+    template <typename F, typename... Args>
+    auto then(stlab::future<void>&& future, F&& f, Args&&... args) {
+        auto task = [_f = std::forward<F>(f), ... _args = std::forward<Args>(args),
+                     _this = this->shared_from_this()]() mutable {
+            temp_thread_name ttn(_this->_name.c_str());
+            if constexpr (std::is_same_v<T, void>) {
+                return std::move(_f)(std::move(_args)...);
+            } else {
+                return std::move(_f)(*(_this->_instance._x), std::move(_args)...);
+            }
+        };
+        return std::move(future).then(_q.executor(), std::move(task));
+    }
+
+    value_instance<T> _instance;
+    stlab::serial_queue_t _q;
+    std::string _name;
+};
+
+/**************************************************************************************************/
+
+} // namespace detail
+
+/**************************************************************************************************/
+
+template <class T>
+class actor {
+    std::shared_ptr<detail::actor_instance<T>> _impl;
+
+public:
+    actor() = default;
+
+    template <class Executor, class... Args>
+    actor(Executor&& e, std::string&& name, Args&&... args) :
```

Comment:
e and name should be values and not r-value references.

Reply:
Fixed.